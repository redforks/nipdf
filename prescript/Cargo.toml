[package]
name = "prescript"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
winnow = { version = "0.5", features = ["simd"] }
log = "0.4"
either = "1.9.0"
memchr = "2.5.0"
ascii85 = "0.2.1"
educe = "0.4.22"
thiserror = "1.0"
anyhow = "1.0"
hex = "0.4"

[dev-dependencies]
test-case = "3.1"
test-log = "0.2"
env_logger = "0.10"
