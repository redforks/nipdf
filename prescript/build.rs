//! Build script that convert `names` file located at workspace root directory,
//! to src/name/built_in_names.rs.
//!
//! Each line in `names` file is a PostScript name, line start with '#' is a comment,
//! empty line is ignored.
//! Sort the names, and generate the builtin_in_names.rs file

use std::{fs::read_to_string, io::Write, path::Path};

fn main() {
    let names_file = Path::join(Path::new(env!("CARGO_MANIFEST_DIR")), "../names");
    let dest_file = Path::join(
        Path::new(env!("CARGO_MANIFEST_DIR")),
        "src/name/built_in_names.rs",
    );

    // return if names_file is not newer than dest_file
    if let Ok(names_meta) = std::fs::metadata(&names_file) {
        if let Ok(dest_meta) = std::fs::metadata(&dest_file) {
            if names_meta.modified().unwrap() <= dest_meta.modified().unwrap() {
                return;
            }
        }
    }

    let names = read_to_string(names_file).unwrap();
    let mut names: Vec<_> = names
        .lines()
        .map(|s| s.trim())
        .filter(|w| !w.is_empty() && !w.starts_with('#'))
        .collect();
    names.sort_unstable();
    names.dedup();
    let mut file = std::fs::File::create(dest_file).unwrap();
    let n = names.len();
    writeln!(file, "pub(crate) static BUILT_IN_NAMES: [&str; {n}] = [").unwrap();
    for name in names {
        writeln!(file, "    {:?},", name).unwrap();
    }
    writeln!(file, "];").unwrap();
}
