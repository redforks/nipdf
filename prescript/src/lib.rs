pub(crate) mod machine;
pub(crate) mod parser;

mod encoding;
mod type1;
pub use encoding::Encoding;
use string_interner::{backend::BucketBackend, symbol::SymbolU16, StringInterner};
pub use type1::Font;

/// Symbol for .notdef glyph
pub const NOTDEF: &str = ".notdef";
pub type Name = SymbolU16;
pub type NameRegistry = StringInterner<BucketBackend<Name>>;
