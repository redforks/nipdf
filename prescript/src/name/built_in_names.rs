#[rustfmt::skip]
pub const BUILT_IN_NAMES: [&str; 7] = ["<<", ">>", "[", "]", "for", "if", ".notdef"];
