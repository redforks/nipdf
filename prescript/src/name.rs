//! Efficient way to store PostScript Name Value
//!
//! PostScript has many frequently used names, such as font glyph names,
//! operation names etc. Use `String`, `Vec<str>`, `Box<str>` needs a lot
//! of allocations, and `Rc<str>` complex code logic.
//!
//! This crate provides a efficient way to store PostScript Name Value.
//! `Name` type wraps a `Either` enum type, which is `Either<u16, Box<str>>`.
//! Predefine common used names, given each name a unique u16 number, and
//! other names stored in Box<str>. Static strings stored in a const array,
//! sorted by alphabetical order, so we can use binary search to find a name.
//!
//! Use `name!()` macro to resolve `&'static str` to Name at compile time,
//! no runtime cost.

use either::Either::{self, Left, Right};

mod built_in_names;
// Hack to use name!() macro
use crate as prescript;
use built_in_names::BUILT_IN_NAMES;
use prescript_macro::name;
use std::fmt::{Display, Formatter};

/// PostScript Name Value
/// Important: Always use name function and macro create Name, don't create
/// Name directly. The internal structure exposed for pattern match combined with `name!` macro:
///
/// match name {
///   name!("foo") => { ... }
///   name!("bar") => { ...}
///   _ => { ... }
/// }  
#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Name(pub Either<u16, Box<str>>);

impl std::fmt::Debug for Name {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "name!({})", self.as_ref())
    }
}

/// Special name to match normally won't exist.
pub static INVALID1: Name = name!("$$invalid1$$");
/// Special name to match normally won't exist.
pub static INVALID2: Name = name!("$$invalid2$$");

impl Display for Name {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "/{}", self.as_ref())
    }
}

impl AsRef<str> for Name {
    fn as_ref(&self) -> &str {
        match &self.0 {
            Left(i) => BUILT_IN_NAMES[*i as usize],
            Right(s) => s.as_ref(),
        }
    }
}

/// Create Name from `&str`, if it is one of builtin names, `Name` will
/// use `u16` to store it, otherwise, `Name` will use `Box<str>` to store it.
///
/// Preferred to use `name!()` macro if possible.
pub fn name(s: &str) -> Name {
    // binary search BUILT_IN_NAMES, return its index if found, otherwise return boxed str
    match BUILT_IN_NAMES.binary_search(&s) {
        Ok(i) => Name(Left(i as u16)),
        Err(_) => Name(Right(Box::from(s))),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use static_assertions::assert_eq_size;

    #[test]
    fn test_name_size() {
        assert_eq_size!(Name, (usize, usize));
    }

    #[test]
    fn name_from_str() {
        assert!(matches!(name("foo"), Name(Right(_))));
        assert!(matches!(name("for"), Name(Left(_))));
    }

    #[test]
    fn as_str() {
        assert_eq!("foo", name("foo").as_ref());
        assert_eq!("for", name("for").as_ref());
    }

    #[test]
    fn name_macro() {
        assert_eq!(name!("foo"), name("foo"));
        assert_eq!(name!("for"), name("for"));
    }

    #[test]
    fn display() {
        assert_eq!(format!("{}", name!("foo")), "/foo");
        assert_eq!(format!("{}", name!("for")), "/for");
    }
}
