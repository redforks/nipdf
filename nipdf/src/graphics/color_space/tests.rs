use super::*;
use crate::{
    file::XRefTable,
    function::{FunctionValue, MockFunction},
};
use assert_approx_eq::assert_approx_eq;
use mockall::predicate::*;
use prescript::name;
use test_case::test_case;
use tinyvec::tiny_vec;

#[test]
fn device_gray_to_rgb() {
    let color_space = DeviceGray;
    let rgba = color_space.to_rgba(&[0x80]);
    assert_eq!(rgba, [0x80, 0x80, 0x80, 0xff]);
}

#[test]
fn rgb_to_rgb() {
    let color_space = DeviceRGB;
    let color = [0x1, 0x2, 0x3];
    let rgba = color_space.to_rgba(&color);
    assert_eq!(rgba, [1, 2, 3, 255]);
}

#[test]
fn cmyk_to_rgb() {
    let color_space = DeviceCMYK;
    let color = [0, 0, 0, 0];
    let rgb = color_space.to_rgba(&color);
    assert_eq!(rgb, [255, 255, 255, 255]);

    let color = [255, 0, 0, 0];
    let rgb = color_space.to_rgba(&color);
    assert_eq!(rgb, [0, 173, 239, 255]);
}

fn to_u8(v: impl ColorCompConvertTo<u8>) -> u8 {
    v.into_color_comp()
}

#[test]
fn convert_color_comp_u8_to_f32() {
    assert_eq!(
        0.0f32,
        <u8 as ColorCompConvertTo<f32>>::into_color_comp(0_u8)
    );
    assert_eq!(
        1.0f32,
        <u8 as ColorCompConvertTo<f32>>::into_color_comp(255_u8)
    );
    assert_approx_eq!(
        0.5f32,
        ColorCompConvertTo::<f32>::into_color_comp(127_u8),
        0.01
    );
}

#[test]
fn convert_color_com_f32_to_u8() {
    assert_eq!(0_u8, to_u8(0.0f32));
    assert_eq!(255_u8, to_u8(1.0f32));
    assert_eq!(128_u8, to_u8(0.5f32)); // round integer part
    assert_eq!(0_u8, ColorCompConvertTo::<u8>::into_color_comp(-1.0f32));
    assert_eq!(255_u8, to_u8(33f32));
}

#[test]
fn test_color_to_rgba() {
    // DeviceGray u8 to u8 rgba
    let color_space = DeviceGray;
    assert_eq!(
        color_to_rgba::<_, u8, _>(&color_space, &[0x80]),
        [0x80, 0x80, 0x80, 0xff]
    );

    // DeviceGray u8 to f32 rgba
    let color_space = DeviceGray;
    assert_eq!(
        color_to_rgba::<_, f32, _>(&color_space, &[51]),
        [0.2f32, 0.2f32, 0.2f32, 1.0f32]
    );
}

#[test]
fn indexed_color_space() {
    let color_space = IndexedColorSpace {
        base: ColorSpace::DeviceRGB,
        data: vec![
            0x00, 0x00, 0x00, // black
            0xff, 0xff, 0xff, // white
        ],
    };
    assert_eq!(2, color_space.len());
    assert_eq!(color_space.to_rgba(&[0]), [0, 0, 0, 255]);
    assert_eq!(color_space.to_rgba(&[1]), [255, 255, 255, 255]);
}

#[test_case("DeviceRGB" => ColorSpace::DeviceRGB)]
#[test_case("DeviceGray" => ColorSpace::DeviceGray)]
#[test_case("DeviceCMYK" => ColorSpace::DeviceCMYK)]
#[test_case("Pattern" => ColorSpace::Pattern(Box::new(PatternColorSpace(None))))]
fn simple_color_space_from_args(nm: &str) -> ColorSpace<f32> {
    let empty_xref = XRefTable::empty();
    let resolver = ObjectResolver::empty(&empty_xref);
    ColorSpace::<f32>::from_args(&ColorSpaceArgs::Name(name(nm)), &resolver, None).unwrap()
}

#[test]
fn icc_based() -> AnyResult<()> {
    // use Alternate color space
    let buf = br#"
1 0 obj
[/ICCBased 2 0 R]
endobj
2 0 obj
<</Length 0/N 1/Alternate /DeviceGray>>
stream
endstream
endobj
"#;
    let xref = XRefTable::from_buf(buf);
    let resolver = ObjectResolver::new(buf, &xref, None);
    let args = ColorSpaceArgs::try_from(resolver.resolve(1)?)?;
    let color_space = ColorSpace::<f32>::from_args(&args, &resolver, None)?;
    assert_eq!(ColorSpace::DeviceGray, color_space);

    // if no Alternate, use Device{Gray, RGB, CMYK} by N value
    for (n, exp) in [
        (1, ColorSpace::DeviceGray),
        (3, ColorSpace::DeviceRGB),
        (4, ColorSpace::DeviceCMYK),
    ] {
        let buf = format!(
            r#"
1 0 obj
[/ICCBased 2 0 R]
endobj
2 0 obj
<</Length 0/N {}>>
stream
endstream
endobj
"#,
            n
        );
        let buf = buf.as_bytes();
        let xref = XRefTable::from_buf(buf);
        let resolver = ObjectResolver::new(buf, &xref, None);
        let args = ColorSpaceArgs::try_from(resolver.resolve(1)?)?;
        let color_space = ColorSpace::<f32>::from_args(&args, &resolver, None)?;
        assert_eq!(exp, color_space);
    }

    Ok(())
}

#[test]
fn separation_color_space() {
    let mut f = MockFunction::new();
    f.expect_call()
        .with(eq(vec![0.5f32]))
        .returning(|_| Ok(tiny_vec![0.1f32, 0.2f32, 0.3f32] as FunctionValue));
    let cs = SeparationColorSpace::<f32> {
        alt: ColorSpace::DeviceRGB,
        f: Rc::new(f),
    };

    assert_eq!(cs.to_rgba(&[0.5f32]), [0.1f32, 0.2f32, 0.3f32, 1.0]);
}

#[test]
fn separation() -> AnyResult<()> {
    // use Alternate color space
    let buf = br#"
1 0 obj
[/Separation /Black /DeviceGray 2 0 R]
endobj
2 0 obj
<</FunctionType 2/Domain [0 1]/N 1>>
endobj
"#;
    let xref = XRefTable::from_buf(buf);
    let resolver = ObjectResolver::new(buf, &xref, None);
    let args = ColorSpaceArgs::try_from(resolver.resolve(1)?)?;
    let color_space = ColorSpace::<f32>::from_args(&args, &resolver, None)?;
    assert_eq!(
        ColorSpace::Separation(Box::new(SeparationColorSpace {
            alt: ColorSpace::DeviceGray,
            f: Rc::new(MockFunction::new())
        })),
        color_space
    );
    Ok(())
}

#[test_case(b"1 0 obj
[/Indexed /DeviceRGB 1 2 0 R]
endobj
2 0 obj
<</Length 6>>
stream
\x01\x02\x03\x04\x05\x06
endstream
endobj
"; "stream")]
#[test_case(b"1 0 obj[/Indexed/DeviceRGB 1(\x01\x02\x03\x04\x05\x06)]endobj"; "Literal String")]
#[test_case(b"1 0 obj[/Indexed/DeviceRGB 1<010203040506>]endobj"; "Hex String")]
fn indexed(buf: &[u8]) -> AnyResult<()> {
    let xref = XRefTable::from_buf(buf);
    let resolver = ObjectResolver::new(buf, &xref, None);
    let args = ColorSpaceArgs::try_from(resolver.resolve(1)?).unwrap();
    let color_space = ColorSpace::<f32>::from_args(&args, &resolver, None).unwrap();
    assert_eq!(
        ColorSpace::Indexed(Box::new(IndexedColorSpace {
            base: ColorSpace::DeviceRGB,
            data: vec![1, 2, 3, 4, 5, 6]
        })),
        color_space
    );
    Ok(())
}

#[test]
fn cal_rgb_from_args() {
    let buf = br#" 
1 0 obj
[/CalRGB <</WhitePoint[0.9505 1.0 1.089]/BlackPoint[0.01 0.02 0.03]/Gamma[1.8 1.8 1.8]/Matrix[0.4497 0.2446 0.0252 0.3163 0.672 0.1412 0.1845 0.0833 0.9227]>>]
endobj
"#;
    let xref = XRefTable::from_buf(buf);
    let resolver = ObjectResolver::new(buf, &xref, None);
    let args = ColorSpaceArgs::try_from(resolver.resolve(1).unwrap()).unwrap();
    let color_space = ColorSpace::<f32>::from_args(&args, &resolver, None).unwrap();
    assert_eq!(
        ColorSpace::CalRGB(Box::new(CalRGBColorSpace {
            white_point: [0.9505, 1.0, 1.089],
            black_point: [0.01, 0.02, 0.03],
            gamma: [1.8, 1.8, 1.8],
            matrix: [
                0.4497, 0.2446, 0.0252, 0.3163, 0.672, 0.1412, 0.1845, 0.0833, 0.9227
            ],
        })),
        color_space
    );
}

#[test]
fn pattern_with_cs_from_args() {
    let buf = br#"1 0 obj
[/Pattern /DeviceRGB]
endobj
"#;
    let xref = XRefTable::from_buf(buf);
    let resolver = ObjectResolver::new(buf, &xref, None);
    let args = ColorSpaceArgs::try_from(resolver.resolve(1).unwrap()).unwrap();
    let color_space = ColorSpace::<f32>::from_args(&args, &resolver, None).unwrap();
    assert_eq!(
        ColorSpace::Pattern(Box::new(PatternColorSpace(Some(ColorSpace::DeviceRGB)))),
        color_space
    );
}

#[test]
fn cal_rgb_color_space() {
    let cs = CalRGBColorSpace {
        white_point: [0.9505, 1.0, 1.089],
        black_point: [0.01, 0.02, 0.03],
        gamma: [1.8, 1.8, 1.8],
        matrix: [
            0.4497, 0.2446, 0.0252, 0.3163, 0.672, 0.1412, 0.1845, 0.0833, 0.9227,
        ],
    };

    assert_eq!(cs.to_rgba(&[0., 1.0, 0.5]), [0., 1.0, 0.5, 1.0]);
}

#[test]
#[should_panic(expected = "Pattern CS base CS not set")]
fn pattern_no_base_panic_to_rgba() {
    let cs = PatternColorSpace::<f32>(None);
    cs.to_rgba(&[1.0, 0., 0., 0.]);
}

#[test]
fn pattern_color_space() {
    // no base color space
    let cs = PatternColorSpace::<f32>(None);
    assert_eq!(0, cs.components());

    // has base color space
    let cs = PatternColorSpace::<f32>(Some(ColorSpace::DeviceRGB));
    assert_eq!(3, cs.components());
    assert_eq!(cs.to_rgba(&[1.0, 0., 0., 0.]), [1.0, 0., 0., 1.]);
}

#[test]
fn device_n_from_args() -> AnyResult<()> {
    let buf = br#"1 0 obj
[/DeviceN [/foo] /DeviceRGB 2 0 R]
endobj
2 0 obj
<</FunctionType 2/Domain [0 1]/N 1>>
endobj
% has attribute
3 0 obj
[/DeviceN [/bar /None] /DeviceRGB 2 0 R <<>>]
endobj
% all color names are None
4 0 obj
[/DeviceN [/None /None] /DeviceRGB 2 0 R]
endobj
"#;
    let xref = XRefTable::from_buf(buf);
    let resolver = ObjectResolver::new(buf, &xref, None);

    let args = ColorSpaceArgs::try_from(resolver.resolve(1)?)?;
    let color_space = ColorSpace::<f32>::from_args(&args, &resolver, None)?;
    assert_eq!(1, color_space.components());
    assert!(matches!(color_space, ColorSpace::DeviceN(_)));

    let args = ColorSpaceArgs::try_from(resolver.resolve(3)?)?;
    let color_space = ColorSpace::<f32>::from_args(&args, &resolver, None)?;
    assert_eq!(2, color_space.components());
    assert!(matches!(color_space, ColorSpace::DeviceN(_)));

    let args = ColorSpaceArgs::try_from(resolver.resolve(4)?)?;
    let color_space = ColorSpace::<f32>::from_args(&args, &resolver, None);
    assert!(color_space.is_err());
    Ok(())
}

#[test]
fn lab_to_rgb() {
    let cs = LabColorSpace {
        white_point: [1.0, 1., 1.],
        ranges: [
            Domain::new(0., 100.),
            Domain::new(-128., 128.),
            Domain::new(-128., 128.),
        ],
        ..Default::default()
    };

    assert_eq!([0., 0., 0., 1.], cs.to_rgba(&[0., 0., 0.]));
    assert_eq!([0., 1., 0., 1.], cs.to_rgba(&[100., -128., 128.]));
    assert_eq!([6, 0, 0, 255], cs.to_rgba(&[0, 128, 128]));
    assert_eq!([0, 255, 0, 255], cs.to_rgba(&[255, 0, 255]));
}
