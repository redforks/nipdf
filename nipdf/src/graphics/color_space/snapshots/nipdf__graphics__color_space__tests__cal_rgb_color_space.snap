---
source: nipdf/src/graphics/color_space/tests.rs
expression: "cs.to_rgba(&[0.4, 0.5, 0.6])"
---
[
    0.16671406,
    0.31006765,
    0.4272814,
    1.0,
]
